lib/telemetry_metrics.ex

```
@@ -1,5 +1,105 @@
+  from event metadata and converted to strings, but you can also provide your own function
+  translating event metadata into a tagset.
   """
+
+  alias Telemetry.Metrics.MetricDefinition
+
+  @type metric_name :: [atom(), ...]
+  @type metric_type :: :counter
+  @type tag_key :: atom()
+  @type tag_value :: String.t()
+  @type tagset :: %{tag_key() => tag_value()}
+  @type datapoint_name :: atom()
+  @type datapoint_value :: number()
+  @type datapoints :: %{datapoint_name => datapoint_value}
+  @type measurement :: %{tagset: tagset(), datapoints: datapoints()}
+  @type option :: {:name, metric_name()} | {:tags, [tag_key()]}
+  @type options :: [option()]
+  @opaque metric_definition :: MetricDefinition.t()
+
+  ## API
+
+  @doc """
+  Returns the metric definition which can be registered in the registry.
+
+  In Telemetry, four values are required to define a metric:
+  * Type which determines what datapoints will be produced;
+  * Event whose values are to be aggregated by the metric;
+  * Name uniquely identifying the metric the scope of a single registry;
+  * Tags which are used to group datapoints
+
+  By default, the name of the metric is the name of the event and the metric is using no tags.
+  `metric_opts` are passed to the metric upon initialization.
+  """
+  @spec new(metric_type(), Telemetry.event_name(), options(), metric_options :: term()) ::
+          metric_definition()
+  def new(metric_type, event_name, opts \\ [], metric_opts \\ []) do
+    metric_name = Keyword.get(opts, :name, event_name)
+    tags = Keyword.get(opts, :tags, [])
+
+    assert_list_of_atoms!(event_name)
+    assert_list_of_atoms!(metric_name)
+    assert_list_of_atoms!(tags)
+
+    MetricDefinition.new(metric_name, metric_type, event_name, tags, metric_opts)
+  end
+
+  ## Helpers
+
+  @spec assert_list_of_atoms!(term()) :: :ok | no_return()
+  defp assert_list_of_atoms!(list) when is_list(list) do
+    Enum.all?(list, &is_atom/1) or raise ArgumentError
```

Let's raise a meaningful error message. :)